crates/ruff_linter/src/source_kind.rs

```diff
@@ -217,11 +217,49 @@ impl<'a> CodeDiff<'a> {
     }
 }
 
+trait ShowNonprinting {
+    fn show_nonprinting(&self) -> String;
+}
+
+impl ShowNonprinting for &str {
+    fn show_nonprinting(&self) -> String {
```

Comment:
It would be nice if we could return a `Cow` here to avoid allocating a string if it doesn't contain any non printable characters. 

We may be able to do something similar to 
https://github.com/mitsuhiko/insta/blob/569bbade9d9ff4bd43fe4138bdcbde00a6bf34c4/insta/src/output.rs#L325-L339

Reply:
Done!